### PR TITLE
Fix proxying of local websocket

### PR DESCRIPTION
**Pull request by costanic**
This fixes an issue where websocket connections are not proxied properly through an upstream proxy configured with -extern-http-proxy-uri.

This can be seen by running edge-proxy with the tunnel-uri feature enabled and an extern-http-proxy-uri configured.  Notice in the netstat output below that edge-proxy has several open connections, ignoring the localhost stuff.  In this output 18.144.154.80 is my upstream HTTP_PROXY specified with extern-http-proxy-uri CLI, yet edge-proxy has connections open to 52.20.116.203 and 34.203.116.53, which are IP addresses associated with gateways.us-east-1.mbedcloud.com, which means that edge-proxy is connecting directly to mbedcloud.com rather than going through the proxy at 18.144.154.80.  After some investigation, it was shown that connection requests with a destination scheme of https were being properly sent to the upstream proxy, but connection requests with a destination scheme of ws were being improperly sent directly to mbedcloud.

This PR fixes the issue by making sure that proxied websocket connections are configured to use the upstream proxy, and by fixing the gorilla/websocket library to be able to connect to https proxies (currently only supports http proxies).

$ sudo netstat -anpt | grep edge-proxy
tcp        0      0 10.0.2.15:34062         18.144.154.80:8443      ESTABLISHED 2785/edge-proxy
tcp        0      0 10.0.2.15:40944         52.20.116.203:443       ESTABLISHED 2785/edge-proxy
tcp        0      0 10.0.2.15:34054         18.144.154.80:8443      ESTABLISHED 2785/edge-proxy
tcp        0      0 10.0.2.15:34066         18.144.154.80:8443      ESTABLISHED 2785/edge-proxy
tcp        0      0 127.0.0.1:51932         127.0.0.1:8080          ESTABLISHED 2785/edge-proxy
tcp        0      0 10.0.2.15:34060         18.144.154.80:8443      ESTABLISHED 2785/edge-proxy
tcp        0      0 10.0.2.15:33222         34.203.116.53:443       ESTABLISHED 2785/edge-proxy
tcp        0      0 10.0.2.15:34080         18.144.154.80:8443      ESTABLISHED 2785/edge-proxy
tcp        0      0 10.0.2.15:34072         18.144.154.80:8443      ESTABLISHED 2785/edge-proxy
tcp        0      0 10.0.2.15:40940         52.20.116.203:443       ESTABLISHED 2785/edge-proxy
tcp        0      0 10.0.2.15:34076         18.144.154.80:8443      ESTABLISHED 2785/edge-proxy
tcp        0      0 10.0.2.15:34056         18.144.154.80:8443      ESTABLISHED 2785/edge-proxy
tcp6       0      0 :::8080                 :::*                    LISTEN      2785/edge-proxy
tcp6       0      0 127.0.0.1:8080          127.0.0.1:51940         ESTABLISHED 2785/edge-proxy
tcp6       0      0 127.0.0.1:8080          127.0.0.1:51964         ESTABLISHED 2785/edge-proxy
tcp6       0      0 127.0.0.1:8080          127.0.0.1:51974         ESTABLISHED 2785/edge-proxy
tcp6       0      0 127.0.0.1:8080          127.0.0.1:51976         ESTABLISHED 2785/edge-proxy
tcp6       0      0 127.0.0.1:8080          127.0.0.1:51932         ESTABLISHED 2785/edge-proxy
tcp6       0      0 127.0.0.1:8080          127.0.0.1:51936         ESTABLISHED 2785/edge-proxy
tcp6       0      0 127.0.0.1:8080          127.0.0.1:51984         ESTABLISHED 2785/edge-proxy
tcp6       0      0 127.0.0.1:8080          127.0.0.1:51962         ESTABLISHED 2785/edge-proxy
tcp6       0      0 127.0.0.1:8080          127.0.0.1:52002         ESTABLISHED 2785/edge-proxy
tcp6       0      0 127.0.0.1:8080          127.0.0.1:51980         ESTABLISHED 2785/edge-proxy
tcp6       0      0 127.0.0.1:8080          127.0.0.1:51960         ESTABLISHED 2785/edge-proxy

### CODE REVIEW

**Comment by costanic** [comment]
@zhucebuliaolongchuan @ygoyal18  Can you help me fix the edge-jenkins error?  It seems that the version of Go used by the CI job isn't downloading go-modules like it does on my dev system.  I'm using Go-1.14.4 which is supposed to be the correct version.

Also, which branch should this PR be raised against?  It looks like 'dev' branch is out of sync with master, so I just made this against master branch.

**Comment by costanic** [comment]
This builds fine in our snapcraft environment and in our linux-pelion-edge environment.  Why is CI broken?

**Comment by cmonr** [comment]
@zhucebuliaolongchuan  @c1728p9 Would y'all be able to answer the branch question?

**Comment by costanic** [comment]
@zhucebuliaolongchuan Now that CI is passing, would you mind reviewing this PR?

**Review by mray190 — APPROVED** [review]

**Comment by cmonr** [comment]
We can also ping @PelionIoT/team-edge_services_admin for a review in case @zhucebuliaolongchuan isn't available/tied up.

File: cmd/edge-proxy/main.go [hunk 1]
```diff
@@ -50,15 +50,15 @@ var certStrategyOptions cmd.OptionMap = cmd.OptionMap{}
...
 	flag.BoolVar(&useL4Proxy, "use-l4-proxy", false, "Use a layer 4 proxy instead of a layer 7 proxy")
 	flag.StringVar(&ca, "ca", "", "Certificate authority for the cloud")
 	flag.StringVar(&certStrategy, "cert-strategy", fog_tls.DefaultDriver(), fmt.Sprintf("Certificate strategy must be one of: %v", fog_tls.Drivers()))
 	flag.Var(&certStrategyOptions, "cert-strategy-options", "Can be specified one or more times. Must be a key-value pair (<key>=<value>)")
-	flag.StringVar(&forwardingAddressesMap, "forwarding-addresses", "{}", "Forwarding address map that proxy server will use to proxy the requests to the corresponding host. Must be a json string")
+	flag.StringVar(&forwardingAddressesMap, "forwarding-addresses", "{}", "Map of local address to forwarded addressfor outgoing HTTP requests. For each forwarding request received at proxy-listen, the destination URI in the request is rewritten based on this map, where the destination server is replaced with the value of the corresponding key.  If the destination server isn't found in this map, then the value of proxy-uri is used.  Must be a json string")
```
Comment (zhucebuliaolongchuan) [inline cmd/edge-proxy/main.go:61]: addressfor -> address for

**Review by zhucebuliaolongchuan — COMMENTED** [review]
@costanic If I am understanding correctly, I think you are trying to make sure all of the websocket connections are being forwarded to the upstream proxy server,  correct me if I am wrong. While we do have websocket connections that connects to gateways domain (gateways.us-east-1.mbedcloud.com) instead of upstream proxy server for reverse tunneling feature, and this behaviour are intended. Actually, this isn't really about the different scheme here, but because of the host of the request sent to edge-proxy and the mapping we have setup for edge-proxy. As you can see from our edge proxy [setup](https://github.com/PelionIoT/snap-pelion-edge/blob/238b19a194a67d19142283ab231d058503a45dfc/files/edge-proxy/scripts/launch-edge-proxy.sh#L51) from snap, parameter `tunnel-uri` has server address `gateways.local`, and it matches one of the keys within `forwarding-addresses`, and that actually take the request with host `gateways.local` to the corresponding upstream server.

Please let me know if you have any other problems on this.

File: cmd/edge-proxy/main.go (hunk 1 above)
Comment (costanic) [reply cmd/edge-proxy/main.go]: fixed, thanks

**Comment by costanic** [comment]
> @costanic If I am understanding correctly, I think you are trying to make sure all of the websocket connections are being forwarded to the upstream proxy server, correct me if I am wrong. While we do have websocket connections that connects to gateways domain (gateways.us-east-1.mbedcloud.com) instead of upstream proxy server for reverse tunneling feature, and this behaviour are intended. Actually, this isn't really about the different scheme here, but because of the host of the request sent to edge-proxy and the mapping we have setup for edge-proxy. As you can see from our edge proxy [setup](https://github.com/PelionIoT/snap-pelion-edge/blob/238b19a194a67d19142283ab231d058503a45dfc/files/edge-proxy/scripts/launch-edge-proxy.sh#L51) from snap, parameter `tunnel-uri` has server address `gateways.local`, and it matches one of the keys within `forwarding-addresses`, and that actually take the request with host `gateways.local` to the corresponding upstream server.
> 
> Please let me know if you have any other problems on this.

Yes, we need to make sure that *all* outbound connections connect through an upstream proxy in order to traverse customer environments with restrictive firewalls where the only way to connect to the internet is through an upstream proxy that sits between the customer's network and the internet.  Without this PR, edge-proxy won't be able to connect directly to mbedcloud.com in such an environment.

**Comment by zhucebuliaolongchuan** [comment]
> > @costanic If I am understanding correctly, I think you are trying to make sure all of the websocket connections are being forwarded to the upstream proxy server, correct me if I am wrong. While we do have websocket connections that connects to gateways domain (gateways.us-east-1.mbedcloud.com) instead of upstream proxy server for reverse tunneling feature, and this behaviour are intended. Actually, this isn't really about the different scheme here, but because of the host of the request sent to edge-proxy and the mapping we have setup for edge-proxy. As you can see from our edge proxy [setup](https://github.com/PelionIoT/snap-pelion-edge/blob/238b19a194a67d19142283ab231d058503a45dfc/files/edge-proxy/scripts/launch-edge-proxy.sh#L51) from snap, parameter `tunnel-uri` has server address `gateways.local`, and it matches one of the keys within `forwarding-addresses`, and that actually take the request with host `gateways.local` to the corresponding upstream server.
> > Please let me know if you have any other problems on this.
> 
> Yes, we need to make sure that _all_ outbound connections connect through an upstream proxy in order to traverse customer environments with restrictive firewalls where the only way to connect to the internet is through an upstream proxy that sits between the customer's network and the internet. Without this PR, edge-proxy won't be able to connect directly to mbedcloud.com in such an environment.

Gotcha.

**Review by zhucebuliaolongchuan — APPROVED** [review]
LGTM

**Comment by costanic** [comment]
Thanks @zhucebuliaolongchuan, are you ok with me merging this to master branch?  On a related note, what should be done with the 2 commits on dev branch?

**Comment by zhucebuliaolongchuan** [comment]
> Thanks @zhucebuliaolongchuan, are you ok with me merging this to master branch? On a related note, what should be done with the 2 commits on dev branch?

Sure. Go ahead. I will merge that change later if I think it needs to.

**Pull request merged**